### DISCUSSION TITLE
[Intel/Haswell/Desktop] Core i5-4570 IMC Timings and DIMM geometry

### DISCUSSION THREAD

**Discussion by cyring** [cyring/CoreFreq#369] [General]
# [`i5-4570`](https://gist.github.com/svmlegacy/729c832d38d879551f8af2050f044a36)
## Timings
Setting | _CoreFreq_ | BIOS
----------|------------------|--------
`tRRD` | `5` | `4`
`tWR` | N/A  | `10`
`tFAW` | N/A  | `20`
`tRTP` | N/A  | `5`
`tWTP` | N/A  | `21`
`drWR` | `4` | `5`
`tCKE` | N/A | `4`

## Geometry
```
( 8 x Bank x Rank x Rows x Cols ) / ( 1024 x 1024 )
```
```
( 8 x 8 x 2 x 16384 x 1024 ) / ( 1024 x 1024) = 2048 MB vs 8192 MB
```

_Originally posted by @cyring in https://github.com/cyring/CoreFreq/discussions/342#discussioncomment-2934706_

**Comment by cyring** [comment]
@svmlegacy and any Broadwell, Haswell owner:

I need to dump Register at MCHBAR address `0x4c00`
(_This register is undocumented, so we are reversing in this [header file](https://github.com/cyring/CoreFreq/blob/2fe50f3d859b479b3b54f178d2a4939019663308/intelmsr.h#L3269)_)

Based on `develop` branch, in function `HSW_IMC()`, add the following `printf` to dump the register value.

https://github.com/cyring/CoreFreq/blob/2fe50f3d859b479b3b54f178d2a4939019663308/corefreqd.c#L3824

```C
    for (cha = 0; cha < RO(Shm)->Uncore.MC[mc].ChannelCount; cha++)
    {
	printf( "Cha(%d) REG4C00[%x]\n", cha,
		RO(Proc)->Uncore.MC[mc].Channel[cha].HSW.REG4C00.value );

```

Please post the output of `corefreqd` here, bellow:

**Comment by cyring** [comment]
@svmlegacy : For IMC testing, here's a Haswell/Broadwell development archive.

[CoreFreq_develop.tar.gz](https://github.com/cyring/CoreFreq/files/9739853/CoreFreq_develop.tar.gz)

**Comment by cyring** [comment]
@svmlegacy Hello,

Changes are now tracked through a dedicated Haswell, Broadwell branch

You just have to pull and build [develop_HSW_BDW](https://github.com/cyring/CoreFreq/tree/develop_HSW_BDW) to test the IMC

Thank you

**Comment by svmlegacy** [comment]
Intel Core i5-4570

Develop branch:
![image](https://user-images.githubusercontent.com/59940364/194941327-1a52b06a-dfc5-4d99-9cf1-c2a2e5b41d10.png)

develop_HSW_BDW branch:
![image](https://user-images.githubusercontent.com/59940364/194941703-786d9507-998b-4301-8191-c946bd53ea0c.png)

**Comment by cyring** [comment]
> develop_HSW_BDW branch: 


Thank you.

Can you pull and test `develop_HSW_BDW` in which I'm trying to better estimate the DIMM slots count.

**Comment by svmlegacy** [comment]
![image](https://user-images.githubusercontent.com/59940364/195204319-720f7816-0d77-480d-bd1e-4e2db689cba7.png)
Motherboard is a Gigabyte GA-Z97-HD3P Rev. 1.0 Four DIMM slots available, slots "3" & "4" populated (Closest and 3rd closest to CPU socket)
![image](https://user-images.githubusercontent.com/59940364/195204671-b7ee897b-070e-4c6f-bed1-14e92269893d.png)

**Comment by cyring** [comment]
> Motherboard is a Gigabyte GA-Z97-HD3P Rev. 1.0 Four DIMM slots available, slots "3" & "4" populated (Closest and 3rd closest to CPU socket) 

If dual channel set, topology looks OK but what is wrong is the number of rows: it should be `65536` instead of `16384`

**Comment by cyring** [comment]
@svmlegacy Can you please pull and test the [`develop`](https://github.com/cyring/CoreFreq/commit/c2aa1ccfd9e43c59b3c5dcb0ad5fac858874e040) branch ?

DIMM rows count is now computed using a formula made for Skylake.

**Comment by cyring** [comment]
> @svmlegacy Can you please pull and test the [`develop`](https://github.com/cyring/CoreFreq/commit/c2aa1ccfd9e43c59b3c5dcb0ad5fac858874e040) branch ?
> 
> DIMM rows count is now computed using a formula made for Skylake.

@svmlegacy could you please test Haswell IMC from latest `develop` ?

**Comment by svmlegacy** [comment]
Fresh pull from Develop:

![image](https://user-images.githubusercontent.com/59940364/195719237-9a6dfe00-b501-414c-abb1-af0a0539cdcd.png)

**Comment by cyring** [comment]
@svmlegacy Can you please pull and try `develop` again ?

Commit attempts to fix the DIMM banks. 
I think it should be `8` for `CM3X8GA2133C9D8`
It can be confirmed if you run `dmidecode -t memory`

**Comment by svmlegacy** [comment]
> @svmlegacy Can you please pull and try `develop` again ?
> 
> Commit attempts to fix the DIMM banks. I think it should be `8` for `CM3X8GA2133C9D8` It can be confirmed if you run `dmidecode -t memory`

```
$ ./corefreq-cli -M
                            Lynx Point-M  [ C00]                           
Controller #0                                                Dual Channel  
 Bus Rate  1333 MHz       Bus Speed 1333 MHz           DDR3 Speed 1333 MHz 
                                                                           
 Cha    CL  RCD   RP  RAS  RRD  RFC   WR RTPr WTPr  FAW  B2B  CWL CMD  REFI
  #0     9    9    9   24    5  174    0    0    0    0    0    7  1T  5200
  #1     9    9    9   24    5  174    0    0    0    0    0    7  1T  5200
      ddWR drWR srWR ddRW drRW srRW ddRR drRR srRR ddWW drWW srWW CKE   ECC
  #0     5    4   18    9    9    9    6    6    4    7    7    4   0    0 
  #1     5    4   18    9    9    9    6    6    4    7    7    4   0    0 
                                                                           
 DIMM Geometry for channel #0                                              
      Slot Bank Rank     Rows   Columns    Memory Size (MB)                
       #0     8    2     65536      1024           8192  CM3X8GA2133C9D8   
       #1                                                                  
 DIMM Geometry for channel #1                                              
      Slot Bank Rank     Rows   Columns    Memory Size (MB)                
       #0     8    2     65536      1024           8192  CM3X8GA2133C9D8   
       #1                                                                  
```

**Comment by svmlegacy** [comment]
Adding my Xeon E3-1270L v4 (Broadwell LGA1150) before I switch back to 1155...

```
$ ./corefreq-cli -M
                           X99/Wellsburg  [1618]                           
Controller #0                                                Dual Channel  
 Bus Rate  2133 MHz       Bus Speed 2128 MHz           DDR3 Speed 2133 MHz 
                                                                           
 Cha    CL  RCD   RP  RAS  RRD  RFC   WR RTPr WTPr  FAW  B2B  CWL CMD  REFI
  #0     9    9    9   24    5  174    0    0    0    0    0    7  1T  5200
  #1     9    9    9   24    5  174    0    0    0    0    0    7  1T  5200
      ddWR drWR srWR ddRW drRW srRW ddRR drRR srRR ddWW drWW srWW CKE   ECC
  #0     5    4   18    9    9    9    6    6    4    7    7    4   0    0 
  #1     5    4   18    9    9    9    6    6    4    7    7    4   0    0 
                                                                           
 DIMM Geometry for channel #0                                              
      Slot Bank Rank     Rows   Columns    Memory Size (MB)                
       #0     8    2     65536      1024           8192  CM3X8GA2133C9D8   
       #1                                                                  
 DIMM Geometry for channel #1                                              
      Slot Bank Rank     Rows   Columns    Memory Size (MB)                
       #0     8    2     65536      1024           8192  CM3X8GA2133C9D8   
       #1                                                                  
```

**Comment by cyring** [comment]
> Adding my Xeon E3-1270L v4 (Broadwell LGA1150) before I switch back to 1155...
> 
> ```
> $ ./corefreq-cli -M
>                            X99/Wellsburg  [1618]                           
> Controller #0                                                Dual Channel  
>  Bus Rate  2133 MHz       Bus Speed 2128 MHz           DDR3 Speed 2133 MHz 
>                                                                            
>  Cha    CL  RCD   RP  RAS  RRD  RFC   WR RTPr WTPr  FAW  B2B  CWL CMD  REFI
>   #0     9    9    9   24    5  174    0    0    0    0    0    7  1T  5200
>   #1     9    9    9   24    5  174    0    0    0    0    0    7  1T  5200
>       ddWR drWR srWR ddRW drRW srRW ddRR drRR srRR ddWW drWW srWW CKE   ECC
>   #0     5    4   18    9    9    9    6    6    4    7    7    4   0    0 
>   #1     5    4   18    9    9    9    6    6    4    7    7    4   0    0 
>                                                                            
>  DIMM Geometry for channel #0                                              
>       Slot Bank Rank     Rows   Columns    Memory Size (MB)                
>        #0     8    2     65536      1024           8192  CM3X8GA2133C9D8   
>        #1                                                                  
>  DIMM Geometry for channel #1                                              
>       Slot Bank Rank     Rows   Columns    Memory Size (MB)                
>        #0     8    2     65536      1024           8192  CM3X8GA2133C9D8   
>        #1                                                                  
> ```

On that Broadwell, `CM3X8GA2133C9D8` DIMMs are this time at `2133 MHz`
Do you confirm it matches BIOS setup ?

**Comment by svmlegacy** [comment]
> On that Broadwell, `CM3X8GA2133C9D8` DIMMs are this time at `2133 MHz`
> Do you confirm it matches BIOS setup ?

BIOS lists memory speed and timings as 1333 MHz, 9-9-9

**Comment by cyring** [comment]
> > On that Broadwell, `CM3X8GA2133C9D8` DIMMs are this time at `2133 MHz`
> > Do you confirm it matches BIOS setup ?
> 
> BIOS lists memory speed and timings as 1333 MHz, 9-9-9

Haswell & Broadwell cases are decoded by same code.
I think that if speed differs, using same DIMM, means both platform registers are not of same bit values.

**Comment by cyring** [comment]
Remaining actions are linked to unspecified Timings registers.

**Comment by cyring** [comment]
Dev ceased
